src/contexts/UserSpellbookContext.js

```diff
@@ -1,15 +1,23 @@
-import { createContext, useState } from "react";
+import { createContext, useReducer, useEffect } from 'react';
+import { bookmarkReducer } from '../reducers/bookmarkReducer';
 
 export const UserSpellbookContext = createContext();
 
 const UserSpellbookContextProvider = (props) => {
-  const [ bookmarks, setBookmarks ] = useState([]);
+  const [ bookmarks, dispatch ] = useReducer(bookmarkReducer, [], () => {
```

Comment:
This is super cool. I feel like using Context together with local storage makes a lot of sense.